packages/legend-data-cube/src/stores/core/DataCubeQuerySnapshotBuilderUtils.ts

```diff
@@ -121,3 +152,385 @@ export function _pruneExpandedPaths(
     )
     .sort();
 }
+
+export function _buildFilterSnapshot(
+  vs: V1_ValueSpecification,
+): DataCubeQuerySnapshotFilter {
+  const filterSnapshot = {} as DataCubeQuerySnapshotFilter;
+  const filterConditionSnapshot = [];
+
+  if (vs instanceof V1_AppliedFunction) {
+    switch (vs.function) {
+      case DataCubeQueryFilterGroupOperator.OR:
+      case DataCubeQueryFilterGroupOperator.AND:
+        filterSnapshot.groupOperator = vs.function;
+        break;
+      default:
+        filterSnapshot.groupOperator = DataCubeQueryFilterGroupOperator.AND;
+    }
+
+    if (
+      vs.function === DataCubeQueryFilterGroupOperator.AND ||
+      vs.function === DataCubeQueryFilterGroupOperator.OR
+    ) {
+      vs.parameters.forEach((param) => {
+        filterConditionSnapshot.push(_buildSubFilter(param)!);
+      });
+    } else {
+      filterConditionSnapshot.push(_buildSubFilter(vs)!);
+    }
+  }
+  filterSnapshot.conditions = filterConditionSnapshot;
+  return filterSnapshot;
+}
+
+function _buildSubFilter(
+  vs: V1_ValueSpecification,
+):
+  | DataCubeQuerySnapshotFilterCondition
+  | DataCubeQuerySnapshotFilter
+  | undefined {
+  if (vs instanceof V1_AppliedFunction) {
+    if (
+      Object.values(DataCubeEngineFilterOperator).includes(
+        vs.function as DataCubeEngineFilterOperator,
+      ) &&
+      vs.function !== DataCubeEngineFilterOperator.NOT
+    ) {
+      return _buildFilterConditionSnapshot(vs);
+    } else if (vs.function === DataCubeEngineFilterOperator.NOT) {
+      const notCondition = _buildNotFilterConditionSnapshot(
+        vs.parameters[0] as V1_AppliedFunction,
+      );
+      if (notCondition) {
+        return notCondition;
+      } else {
+        const filterSnapshot = _buildFilterSnapshot(
+          vs.parameters[0] as V1_ValueSpecification,
+        );
+        filterSnapshot.not = true;
+        return filterSnapshot;
+      }
+    } else if (vs.parameters[0] instanceof V1_AppliedFunction) {
+      return _buildFilterSnapshot(vs);
+    }
+  }
+  return undefined;
+}
+
+function _buildFilterConditionSnapshot(
```

Comment:
as discussed, distribute this logic to each of the operator. For common logic that can be shared, we can leave them here, e.g. _not(), etc.

Reply:
so this is the code for conditionExpression:
const filterCondition = filter;
    const operation = filterOperations.find(
      (op) => op.operator === filterCondition.operator,
    );
    const condition = operation?.buildConditionExpression(filterCondition);
    
    operator for most cases is the same but different for some cases so will need switch case still to put in the operator  but I get the comment